lib/src/inline_syntaxes/footnote_ref_syntax.dart

```diff
@@ -0,0 +1,32 @@
+import '../ast.dart';
+import '../charcode.dart';
+import '../inline_parser.dart';
+import 'inline_syntax.dart';
+
+class FootnoteRefSyntax extends InlineSyntax {
+  FootnoteRefSyntax() : super(r'!?\[\^(\S+?)\]');
+
+  @override
+  bool onMatch(InlineParser parser, Match match) {
+    final label = match[1]!;
+    final refs = parser.document.footnoteReferences;
+    if (!refs.containsKey(label)) {
+      final start = parser.start;
+      final text = parser.source.substring(start, start + match[0]!.length);
+      parser.addNode(Text(text));
+      return true;
+    }
+    refs[label] = (refs[label] ?? 0) + 1;
+    final pos = refs.keys.toList(growable: false).indexOf(label);
+
+    if (parser.charAt(parser.start) == $exclamation) {
```

Comment:
Why is an exclamation handled separately here?

Reply:
code  refined, and not need any more.